omp/matrix/fft_kernels.cpp

```diff
@@ -0,0 +1,357 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/fft_kernels.hpp"
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The FFT matrix format namespace.
+ * @ref Fft
+ * @ingroup fft
+ */
+namespace fft {
+
+
+template <typename InValueType, typename OutValueType>
+void bfly(const matrix::Dense<InValueType> *b, matrix::Dense<OutValueType> *x,
+          int64 lo, int64 hi, OutValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = b->at(lo, rhs);
+        auto hi_val = b->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
+    }
+}
+
+
+template <typename ValueType>
+void bfly(matrix::Dense<ValueType> *x, int64 lo, int64 hi, ValueType root)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        auto lo_val = x->at(lo, rhs);
+        auto hi_val = x->at(hi, rhs);
+        x->at(lo, rhs) = lo_val + hi_val;
+        x->at(hi, rhs) = (lo_val - hi_val) * root;
+    }
+}
+
+
+template <typename ValueType>
+void bit_rev_swap(matrix::Dense<ValueType> *x, int64 i, int64 rev_i)
+{
+    for (size_type rhs = 0; rhs < x->get_size()[1]; rhs++) {
+        if (i < rev_i) {
+            std::swap(x->at(i, rhs), x->at(rev_i, rhs));
+        }
+    }
+}
+
+
+int64 bit_rev(int64 i, int64 size)
+{
+    int64 rev{};
+    for (int fwd = 1, bwd = size / 2; fwd < size; fwd *= 2, bwd /= 2) {
+        rev |= ((i / fwd) & 1) * bwd;
+    }
+    return rev;
+}
+
+
+template <typename ValueType>
+vector<ValueType> build_unit_roots(std::shared_ptr<const DefaultExecutor> exec,
+                                   int64 size, int64 sign)
+{
+    vector<ValueType> roots(size / 2, {exec});
+    for (int64 i = 0; i < size / 2; i++) {
+        roots[i] = unit_root<ValueType>(size, sign * i);
+    }
+    return roots;
+}
+
+
+template <typename InValueType, typename OutValueType>
+void fft_impl(std::shared_ptr<const DefaultExecutor> exec,
+              const matrix::Dense<InValueType> *b,
+              matrix::Dense<OutValueType> *x, bool inverse)
+{
+    using complex_type = OutValueType;
+    using real_type = remove_complex<complex_type>;
+    int64 sign = inverse ? 1 : -1;
+    auto nrhs = b->get_size()[1];
+    auto size = static_cast<int64>(b->get_size()[0]);
```

Comment:
Please make constants `const`
```suggestion
    const int64 sign = inverse ? 1 : -1;
    const auto nrhs = b->get_size()[1];
    const auto size = static_cast<int64>(b->get_size()[0]);
```